### PR TITLE
Work on `ConsumerController` deadlock

### LINKED ISSUE #7088: ConsumerController is Deadlocked in WaitingForConfirmation Status
**Version Information**
Version of Akka.NET? 1.5.15
Which Akka.NET Modules? Akka.Delivery and Akka.Persistence

**Describe the bug**
If the ConsumerController tries to delivery a SequencedMessage<T> to the consumer, but the consumer is not able to be created and doesn't receive the SequencedMessage<T>, then the ConsumerController is deadlocked because it doesn't try to resend the SequencedMessage<T> again. The ProducerController sends a Retry message to the ConsumerController, but the ConsumerController ignores it.

**To Reproduce**
For our issue specifically, this occurs whenever our true "receiving" actor is created by a Manager. The Manager receives the message from the ConsumerController and will forward said message to a child, who is the true consumer. The child is not able to be created/hydrated due to some SQL Timeout or transient error at the time. The ConsumerController was able to "Tell" the Sequenced message but our manager was not able to forward it on. The ConsumerController is now stuck because it doesn't retry to send this message. 

I can create a small POC in Github if needed.

**Expected behavior**
I expect our consumer receive multiple instances of the message, so that the consumer actor has multiple attempts to be created. 

**Actual behavior**
The ConsumerController is stuck in a WaitingForConfirmation status without a way to get out of that status.

**Screenshots**

**Environment**
Azure Kubernetes

**Additional context**

### PR DESCRIPTION
**Pull request by Aaronontheweb**
## Changes

Fixes #7088 - work in progress.

## Checklist

For significant changes, please ensure that the following have been completed (delete if not relevant):

* [x] This change follows the [Akka.NET API Compatibility Guidelines](https://getakka.net/community/contributing/api-changes-compatibility.html).
* [x] This change follows the [Akka.NET Wire Compatibility Guidelines](https://getakka.net/community/contributing/wire-compatibility.html).
* [x] I have [reviewed my own pull request](https://getakka.net/community/contributing/index.html#review-your-own-pull-requests).
* [x] Design discussion issue #7088
* [x] Changes in public API reviewed, if any.
* [ ] I have added website documentation for this feature.

### Latest `dev` Benchmarks 

Include data from the [relevant benchmark](https://getakka.net/community/contributing/index.html#improve-performance) prior to this change here.

### This PR's Benchmarks

Include data from after this change here.

### CODE REVIEW

File: src/core/Akka/Delivery/ConsumerController.cs [hunk 1]
```diff
@@ -229,26 +229,57 @@ public static Settings Create(ActorSystem actorSystem)
...
+        /// <see cref="ProducerController"/> until they have been confirmed, but the drawback is that lost messages
+        /// will not be delivered.
+        /// </summary>
         public bool OnlyFlowControl { get; init; }
+        
+        /// <summary>
+        /// When disabled, the <see cref="ConsumerController"/> will discard any <c>Retry</c> messages when it is
+        /// waiting for a message confirmation.
+        /// </summary>
+        public bool RetryConfirmation { get; init; }
```
Comment (Arkatufus) [inline src/core/Akka/Delivery/ConsumerController.cs:284]: Added new `RetryConfirmation` settings

File: src/core/Akka/Delivery/Internal/ConsumerControllerImpl.cs [hunk 2]
```diff
@@ -386,6 +386,17 @@ long ComputeNextSeqNr()
         Receive<Retry>(_ =>
         {
             // no retries when WaitingForConfirmation, will be performed from (idle) active
+            if (!Settings.RetryConfirmation)
+                return;
```
Comment (Arkatufus) [inline src/core/Akka/Delivery/Internal/ConsumerControllerImpl.cs:390]: Only retry the delivery if `RetryConfirmation` is set to true

File: src/core/Akka/Delivery/Internal/ConsumerControllerImpl.cs [hunk 3]
```diff
@@ -386,6 +386,17 @@ long ComputeNextSeqNr()
...
+            
+            _log.Debug("Consumer received Retry while waiting for confirmation.");
+            ReceiveRetry(() =>
+            {
+                _log.Debug("Retry sending SequencedMessage [{0}].", sequencedMessage.SeqNr);
+                CurrentState.Consumer.Tell(new Delivery<T>(sequencedMessage.Message.Message!, Context.Self,
+                    sequencedMessage.ProducerId, sequencedMessage.SeqNr));
+                Become(() => WaitingForConfirmation(sequencedMessage));
+            });
```
Comment (Arkatufus) [inline src/core/Akka/Delivery/Internal/ConsumerControllerImpl.cs:399]: Retry logic

**Review by Arkatufus — COMMENTED** [review]
Self-review

File: src/core/Akka/Configuration/akka.conf [hunk 4]
```diff
@@ -664,6 +664,8 @@ akka {
       # kept in memory in the `ProducerController` until they have been
       # confirmed, but the drawback is that lost messages will not be delivered.
       only-flow-control = false
+      
+      retry-confirmation = false
```
Comment (Arkatufus) [inline src/core/Akka/Configuration/akka.conf:672]: `RetryConfirmation` is false by default

Comment (Aaronontheweb) [reply src/core/Akka/Configuration/akka.conf:672]: LGTM

File: src/core/Akka.Tests/Delivery/ConsumerControllerSpecs.cs [hunk 5]
```diff
@@ -672,4 +673,26 @@ public async Task ConsumerController_can_process_zero_length_Chunk()
         consumerController.Tell(seqMessages1.First());
         (await consumerProbe.ExpectMsgAsync<ConsumerController.Delivery<ZeroLengthSerializer.TestMsg>>()).Message.Should().Be(ZeroLengthSerializer.TestMsg.Instance);
     }
+    
+    [Fact]
+    public async Task ConsumerController_must_not_resend_Delivery()
```
Comment (Aaronontheweb) [inline src/core/Akka.Tests/Delivery/ConsumerControllerSpecs.cs:678]: LGTM

File: src/core/Akka.Tests/Delivery/ConsumerControllerRetryConfirmSpecs.cs [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+using static Akka.Tests.Delivery.TestConsumer;
+
+namespace Akka.Tests.Delivery;
+
+public class ConsumerControllerRetryConfirmSpecs : TestKit.Xunit2.TestKit
+{
+    public static readonly Config Config = @"
+        akka.reliable-delivery.consumer-controller {
+        flow-control-window = 20
+        resend-interval-min = 1s
```
Comment (Aaronontheweb) [inline src/core/Akka.Tests/Delivery/ConsumerControllerRetryConfirmSpecs.cs]: Might want to lower this given that you're waiting for up to 1.5 seconds - just to provide a bigger window in the event of racy tests

File: src/core/Akka/Delivery/ConsumerController.cs (hunk 1 above)
Comment (Aaronontheweb) [reply src/core/Akka/Delivery/ConsumerController.cs:284]: LGTM

File: src/core/Akka/Delivery/Internal/ConsumerControllerImpl.cs [hunk 7]
```diff
@@ -386,6 +386,17 @@ long ComputeNextSeqNr()
...
+            if (!Settings.RetryConfirmation)
+                return;
+            
+            _log.Debug("Consumer received Retry while waiting for confirmation.");
+            ReceiveRetry(() =>
+            {
+                _log.Debug("Retry sending SequencedMessage [{0}].", sequencedMessage.SeqNr);
+                CurrentState.Consumer.Tell(new Delivery<T>(sequencedMessage.Message.Message!, Context.Self,
+                    sequencedMessage.ProducerId, sequencedMessage.SeqNr));
+                Become(() => WaitingForConfirmation(sequencedMessage));
```
Comment (Aaronontheweb) [inline src/core/Akka/Delivery/Internal/ConsumerControllerImpl.cs:398]: Do we need to do a `Become` again here? Aren't we already in this state?

**Review by Aaronontheweb — COMMENTED** [review]
LGTM but left some questions

File: src/core/Akka/Delivery/Internal/ConsumerControllerImpl.cs (hunk 7 above)
Comment (Arkatufus) [reply src/core/Akka/Delivery/Internal/ConsumerControllerImpl.cs:398]: There's no harm in doing it, we might as well have the code to explicitly show which state we're transitioning to.

File: src/core/Akka.Tests/Delivery/ConsumerControllerRetryConfirmSpecs.cs (hunk 6 above)
Comment (Arkatufus) [reply src/core/Akka.Tests/Delivery/ConsumerControllerRetryConfirmSpecs.cs]: ok, will do

**Comment by Aaronontheweb** [comment]
@Arkatufus merged - we'll need a separate PR to add some documentation for this to the Akka.Delivery page on the website

**Pull request merged**